src/main/cpp/src/hive_hash.cu

```diff
@@ -175,17 +191,25 @@ class hive_device_row_hasher {
    * @param row_index The row index to compute the hash value of
    * @return The hash value of the row
    */
-  __device__ auto operator()(cudf::size_type row_index) const noexcept
+  __device__ hive_hash_value_t operator()(cudf::size_type row_index) const noexcept
   {
+    auto const hash_adapter = element_hasher_adapter{_check_nulls, *this};
     return cudf::detail::accumulate(
-      _table.begin(),
-      _table.end(),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(_num_columns),
       HIVE_INIT_HASH,
       cuda::proclaim_return_type<hive_hash_value_t>(
-        [row_index, nulls = this->_check_nulls] __device__(auto hash, auto const& column) {
-          auto cur_hash =
-            cudf::type_dispatcher(column.type(), element_hasher_adapter{nulls}, column, row_index);
-          return HIVE_HASH_FACTOR * hash + cur_hash;
+        [=] __device__(auto const hash, auto const col_index) {
+          auto const col_info = _col_infos[_column_map[col_index]];
+          auto const col_hash =
+            (col_info.type_id == cudf::type_id::LIST || col_info.type_id == cudf::type_id::STRUCT)
+              ? hash_adapter.hash_nested(col_index, row_index)
```

Comment:
Nit: pass in the index of `_col_infos`, because `_column_map[col_index]` already calculates.
In the `hash_nested`, it's calculated again:
```cpp
auto const flattened_idx = _parent._column_map[col_index];
```